draft-ietf-mls-protocol.md

```diff
@@ -655,80 +584,57 @@ public-key encryption to the public key K of the secret value S):
 | pk(B)      |             |
 
 
-### Blank Ratchet Tree Nodes
-
-Nodes in a ratchet tree can have a special value "\_", used to indicate that the
-node should be ignored during path computations. Such nodes are used to replace
-leaves when participants are deleted from the group.
+# Group State
 
-If any node in the copath of a leaf is \_, it should be ignored during the
-computation of the path. For example, the tree consisting of the private
-keys (A, \_, C, D) is constructed as follows for ART:
+Each participant in the group maintains a representation of the
+state of the group:
 
 ~~~~~
-  DH(A, DH(CD))
-   /      \
-  A       DH(CD)
- / \      /  \
-A   _    C    D
+struct GroupState {
+  opaque group_id<0..255>;
+  uint32 size;
+  uint32 epoch;
+  Credential roster<1..2^24-1>;
+  PublicKey tree<1..2^24-1>;
+  opaque transcript_hash<0..255>;
+}
 ~~~~~
 
-Replacing a node by \_ in TreeKEM, means performing an update on any leaf
-without sending the new key to the the blanked leaf.
-In the following example, participant A update its key to A' and derive the new
-sequence of keys up-to the path. Here A only send H(H(A')) to the parent
-node of C and D but does not send H(A') to B which evicts it from the Group.
-
-~~~
-    H(H(A'))
-    /    \
- H(A')    H(C)
-  / \    /  \
- A'  _  C    D
-~~~
-
-If two sibling nodes are both \_, their parent value also becomes \_.
-
-Blank nodes effectively result in an unbalanced tree, but allow the
-tree management to behave as for a balanced tree for programming simplicity.
-
+The fields in this state have the following semantics:
+
+* The `group_id` field is an application-defined identifier for the
+  group.
+* The `size` field represents the total number of key-exchange slots
+  in the group, whether or not these slots are currently occupied.
+* The `epoch` field represents the current version of the group key.
+* The `roster` field contains credentials for the occupied slots in
+  the tree, including the identity and public key for the holder of
```

Comment:
"Public key" == "Signature public key".  ("Identity" is whatever opaque string the application assigns)  Note that because this state can change, we're not committed to using long-term identity keys.